Libs/GObject/Classes/Object.cs

```diff
@@ -388,38 +333,21 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
-        protected void Dispose(bool disposing)
-        {
-            if (Disposed)
-                return;
-
-            if(disposing)
-                DisposeManagedState();
-
-            DisposeUnmanagedState();
-            SetDisposed();
-        }
-
-        protected void SetDisposed()
+        protected virtual void Dispose(bool disposing)
         {
-            Disposed = true;
-        }
-
-        protected virtual void DisposeManagedState()
-        {
-            Objects.Remove(Handle);
-            Handle = IntPtr.Zero;
+            if (Handle != IntPtr.Zero)
+            {
+                foreach (var signalHelper in _signals.Values)
+                    signalHelper.Dispose();
```

Comment:
Hm, good point. I'm not 100% sure about disposing managed / unmanaged state with / without subclasses. From a documentation point of view I think you are right.

I wonder what the problem would be if we keep it like this? The `signalhelper` could probably be already disposed via its own finalizer right?

I think the dispose method for `Object.ClosureHelper` is fine as it is purely unmanaged code. But I'm not sure about `Object.SignalHelper` anymore as it stores a reference to the `Object.ClosureHelper` which itself is a managed object with it's own dispose mehtod. Probably it would need to change, too?

Reply:
I just asked this on gitter/CSharplang and we should not touch this like you suggested. I got the recomendation to use `SafeHandle` instead of a destructor as it can prevent some race conditions if the pointer is freed during a p/invoke call.

So I'm going to rework this to use finalizers and I'm going to open an issue to port over to `SafeHandles`.

Here some sample links from the discussion:
- https://github.com/Techsola/InstantReplay/blob/main/src/Techsola.InstantReplay/Native/Gdi32.cs#L50
- https://github.com/Techsola/InstantReplay/blob/main/src/Techsola.InstantReplay/Native/Gdi32.GdiObjectSafeHandle.cs
- https://github.com/Techsola/InstantReplay/blob/77862658f3bd5c1a50c048095768150d268bf308/src/Techsola.InstantReplay/Native/Gdi32.cs#L104

If we use safe handles for our handles, we can omit Destructors and the "GC.SupressFinalize" completely as the handles are like managed code which get freed by the `SafeHandle` class:
- https://github.com/dotnet/runtime/discussions/41493#discussioncomment-58355

```cs
class Foo : IDisposable
{
    private FooSafeHandle thing;

    public void Dispose() => thing.Dispose();
}
``